flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_Boarding.js

```diff
@@ -86,139 +91,241 @@ class A32NX_Boarding {
 
-        let isAllCargoStationFilled = true;
-        for (const _station of Object.values(this.cargoStations)) {
-            const stationCurrentLoad = SimVar.GetSimVarValue(`L:${_station.simVar}`, "Number");
-            const stationCurrentLoadTarget = SimVar.GetSimVarValue(`L:${_station.simVar}_DESIRED`, "Number");
+            // For some reason, GSX doesn't emit 100% cargo boarding percentage, so this is required
+            if (gsxDeBoardState === 6) {
+                for (const loadStation of Object.values(this.cargoStations)) {
+                    this.fillCargoStation(loadStation, 0);
+                }
+            }
+
+            if (gsxDeBoardState === 5) {
+                const gsxDeBoardingTotal = SimVar.GetSimVarValue("L:FSDT_GSX_NUMPASSENGERS_DEBOARDING_TOTAL", "Number");
+                this.passengersLeftToFillOrEmpty = gsxDeBoardingTotal - this.prevBoardedOrDeboarded;
+
+                for (const paxStation of Object.values(this.paxStations).reverse()) {
+                    const stationCurrentPax = SimVar.GetSimVarValue(`L:${paxStation.simVar}`, "Number");
+                    const stationCurrentPaxTarget = SimVar.GetSimVarValue(`L:${paxStation.simVar}_DESIRED`, "Number");
+                    if (this.passengersLeftToFillOrEmpty <= 0) {
+                        break;
+                    }
+
+                    if (stationCurrentPax > stationCurrentPaxTarget) {
+                        this.fillPaxStation(paxStation, stationCurrentPax - Math.min(this.passengersLeftToFillOrEmpty, paxStation.seats));
+                        this.passengersLeftToFillOrEmpty -= Math.min(this.passengersLeftToFillOrEmpty, paxStation.seats);
+                    }
+                }
+                this.prevBoardedOrDeboarded = gsxDeBoardingTotal;
 
-            if (stationCurrentLoad !== stationCurrentLoadTarget) {
-                isAllCargoStationFilled = false;
-                break;
+                const gsxCargoDeBoardPercentage = SimVar.GetSimVarValue("L:FSDT_GSX_DEBOARDING_CARGO_PERCENT", "Number");
+                for (const loadStation of Object.values(this.cargoStations)) {
+                    if (this.prevCargoDeboardedPercentage == gsxCargoDeBoardPercentage) {
+                        break;
+                    }
+                    const stationCurrentLoad = SimVar.GetSimVarValue(`L:${loadStation.simVar}`, "Number");
+
+                    const loadAmount = stationCurrentLoad * ((100 - gsxCargoDeBoardPercentage) / 100);
+                    this.fillCargoStation(loadStation, loadAmount);
+                }
+                this.prevCargoDeboardedPercentage = gsxCargoDeBoardPercentage;
             }
-        }
 
-        // Sound Controllers
-        if ((currentPax < paxTarget) && boardingStartedByUser == true) {
-            await SimVar.SetSimVarValue("L:A32NX_SOUND_PAX_BOARDING", "Bool", true);
-            this.isBoarding = true;
+            this.loadPaxPayload();
+            this.loadCargoPayload();
+
```

Comment:
Tbh, while I was working on rebasing my PR, I'm not really sure I am keen on this design where we are essentially duplicating (with a if else) the logic driving boarding stations.

There should be a way to abstract this such that there wouldn't need to be repeated/duplicate code.

Reply:
The problem is the boarding logic isn't really duplicated, each for loop different between GSX and native. Then there's the problem of when those for loops are called. For native it's at the bottom after checking various things.

I can attempt to abstract this to a method that handles GSX vs native loading to reduce duplication, but the original if else will most likely remain

Reply:
resolved [3c420ef](https://github.com/flybywiresim/a32nx/pull/7695/commits/3c420ef6b7e05dbdafcd7d1d457722de51d7a24b)